File: frontend/src/components/benchmark/content/avl_tree/implementations/avl_tree.cs
using System;
using System.Collections.Generic;

public class AVLTree
{
    private class Node
    {
        public int Key;
        public Node Left;
        public Node Right;
        public int Height;

        public Node(int key)
        {
            Key = key;
            Height = 1;
        }
    }

    private Node root;

    private int Height(Node node)
    {
        return node == null ? 0 : node.Height;
    }

    private int Balance(Node node)
    {
        return node == null ? 0 : Height(node.Left) - Height(node.Right);
    }

    private void UpdateHeight(Node node)
    {
        node.Height = 1 + Math.Max(Height(node.Left), Height(node.Right));
    }

    private Node RotateRight(Node y)
    {
        Node x = y.Left;
        Node t2 = x.Right;
        x.Right = y;
        y.Left = t2;
        UpdateHeight(y);
        UpdateHeight(x);
        return x;
    }

    private Node RotateLeft(Node x)
    {
        Node y = x.Right;
        Node t2 = y.Left;
        y.Left = x;
        x.Right = t2;
        UpdateHeight(x);
        UpdateHeight(y);
        return y;
    }

    public void Insert(int key)
    {
        root = Insert(root, key);
    }

    private Node Insert(Node node, int key)
    {
        if (node == null)
            return new Node(key);

        if (key < node.Key)
            node.Left = Insert(node.Left, key);
        else if (key > node.Key)
            node.Right = Insert(node.Right, key);
        else
            return node;

        UpdateHeight(node);
        int balance = Balance(node);

        if (balance > 1 && key < node.Left.Key)
            return RotateRight(node);

        if (balance < -1 && key > node.Right.Key)
            return RotateLeft(node);

        if (balance > 1 && key > node.Left.Key)
        {
            node.Left = RotateLeft(node.Left);
            return RotateRight(node);
        }

        if (balance < -1 && key < node.Right.Key)
        {
            node.Right = RotateRight(node.Right);
            return RotateLeft(node);
        }

        return node;
    }

    public void Delete(int key)
    {
        root = Delete(root, key);
    }

    private Node Delete(Node node, int key)
    {
        if (node == null)
            return null;

        if (key < node.Key)
            node.Left = Delete(node.Left, key);
        else if (key > node.Key)
            node.Right = Delete(node.Right, key);
        else
        {
            if (node.Left == null)
                return node.Right;
            else if (node.Right == null)
                return node.Left;
            else
            {
                Node minNode = FindMin(node.Right);
                node.Key = minNode.Key;
                node.Right = Delete(node.Right, minNode.Key);
            }
        }

        UpdateHeight(node);
        int balance = Balance(node);

        if (balance > 1 && Balance(node.Left) >= 0)
            return RotateRight(node);

        if (balance > 1 && Balance(node.Left) < 0)
        {
            node.Left = RotateLeft(node.Left);
            return RotateRight(node);
        }

        if (balance < -1 && Balance(node.Right) <= 0)
            return RotateLeft(node);

        if (balance < -1 && Balance(node.Right) > 0)
        {
            node.Right = RotateRight(node.Right);
            return RotateLeft(node);
        }

        return node;
    }

    private Node FindMin(Node node)
    {
        Node current = node;
        while (current.Left != null)
            current = current.Left;
        return current;
    }

    public bool Search(int key)
    {
        return Search(root, key);
    }

    private bool Search(Node node, int key)
    {
        if (node == null)
            return false;
        if (key == node.Key)
            return true;
        if (key < node.Key)
            return Search(node.Left, key);
        return Search(node.Right, key);
    }

    public List<int> Inorder()
    {
        List<int> result = new List<int>();
        Inorder(root, result);
        return result;
    }

    private void Inorder(Node node, List<int> result)
    {
        if (node != null)
        {
            Inorder(node.Left, result);
            result.Add(node.Key);
            Inorder(node.Right, result);
        }
    }
}

File: frontend/src/components/benchmark/content/avl_tree/implementations/avl_tree.js
class AVLNode {
    constructor(key) {
        this.key = key;
        this.left = null;
        this.right = null;
        this.height = 1;
    }
}

class AVLTree {
    constructor() {
        this.root = null;
    }

    height(node) {
        return node ? node.height : 0;
    }

    balance(node) {
        return node ? this.height(node.left) - this.height(node.right) : 0;
    }

    updateHeight(node) {
        node.height = 1 + Math.max(this.height(node.left), this.height(node.right));
    }

    rotateRight(y) {
        const x = y.left;
        const t2 = x.right;
        x.right = y;
        y.left = t2;
        this.updateHeight(y);
        this.updateHeight(x);
        return x;
    }

    rotateLeft(x) {
        const y = x.right;
        const t2 = y.left;
        y.left = x;
        x.right = t2;
        this.updateHeight(x);
        this.updateHeight(y);
        return y;
    }

    insert(key) {
        this.root = this._insert(this.root, key);
    }

    _insert(node, key) {
        if (!node) {
            return new AVLNode(key);
        }

        if (key < node.key) {
            node.left = this._insert(node.left, key);
        } else if (key > node.key) {
            node.right = this._insert(node.right, key);
        } else {
            return node;
        }

        this.updateHeight(node);
        const bal = this.balance(node);

        if (bal > 1 && key < node.left.key) {
            return this.rotateRight(node);
        }

        if (bal < -1 && key > node.right.key) {
            return this.rotateLeft(node);
        }

        if (bal > 1 && key > node.left.key) {
            node.left = this.rotateLeft(node.left);
            return this.rotateRight(node);
        }

        if (bal < -1 && key < node.right.key) {
            node.right = this.rotateRight(node.right);
            return this.rotateLeft(node);
        }

        return node;
    }

    delete(key) {
        this.root = this._delete(this.root, key);
    }

    _delete(node, key) {
        if (!node) {
            return null;
        }

        if (key < node.key) {
            node.left = this._delete(node.left, key);
        } else if (key > node.key) {
            node.right = this._delete(node.right, key);
        } else {
            if (!node.left) {
                return node.right;
            } else if (!node.right) {
                return node.left;
            } else {
                const minNode = this._findMin(node.right);
                node.key = minNode.key;
                node.right = this._delete(node.right, minNode.key);
            }
        }

        this.updateHeight(node);
        const bal = this.balance(node);

        if (bal > 1 && this.balance(node.left) >= 0) {
            return this.rotateRight(node);
        }

        if (bal > 1 && this.balance(node.left) < 0) {
            node.left = this.rotateLeft(node.left);
            return this.rotateRight(node);
        }

        if (bal < -1 && this.balance(node.right) <= 0) {
            return this.rotateLeft(node);
        }

        if (bal < -1 && this.balance(node.right) > 0) {
            node.right = this.rotateRight(node.right);
            return this.rotateLeft(node);
        }

        return node;
    }

    _findMin(node) {
        let current = node;
        while (current.left) {
            current = current.left;
        }
        return current;
    }

    search(key) {
        return this._search(this.root, key);
    }

    _search(node, key) {
        if (!node) {
            return false;
        }
        if (key === node.key) {
            return true;
        }
        if (key < node.key) {
            return this._search(node.left, key);
        }
        return this._search(node.right, key);
    }

    inorder() {
        const result = [];
        this._inorder(this.root, result);
        return result;
    }

    _inorder(node, result) {
        if (node) {
            this._inorder(node.left, result);
            result.push(node.key);
            this._inorder(node.right, result);
        }
    }
}

module.exports = AVLTree;